Getting and Cleaning Data
====

### 01 - Introduction

Lending Club is a peer to peer lending company based in the United States, in which investors provide funds for potential borrowers and investors earn a profit depending on the risk they take (the borrowers credit score). Lending Club provides the "bridge" between investors and borrowers.

For more information about the company, please check out the official [website](www.lendingclub.com)

__About the data__

Lending club provides several [csv files](https://www.lendingclub.com/info/download-data.action) that contain complete loan data for all loans issued from 2007 to last 2017 quarter, including the current loan status (Current, Late, Fully Paid, etc.) and latest payment information. Some additional features such as credit scores, number of finance inquiries, zip codes, states, or collections have been included among others.

Lending club provides a Data Dictionary that includes definitions for all the data attributes included in the Historical data file and the In Funding data file.

In this notebook we are going to get familiar with the data following the below plan of action:
* Explore files using the shell.
* Concatenate all files in just one dataset.
* Check up NA's in the dataset and decide what to do with them.
* Select columns of interest using the Lending Club Data Dictionary in XLM format.

Let's get started!!!

### 02 - Explore Files

__Libraries__

In [2]:
#remove warnigns
import warnings
warnings.filterwarnings("ignore")

import os

import numpy as np
import pandas as pd

We are going to use the shell for a first data exploration and getting familiar with the data:

In [6]:
raw_data_path = "../data/raw"

In [8]:
!ls -lh {raw_data_path}

total 1,4G
-rw-rw-r-- 1 juanan juanan 104M may  2 12:24 LoanStats_2016Q1.csv
-rw-rw-r-- 1 juanan juanan  76M may  2 12:42 LoanStats_2016Q2.csv
-rw-rw-r-- 1 juanan juanan  77M may  2 12:56 LoanStats_2016Q3.csv
-rw-rw-r-- 1 juanan juanan  81M may  2 13:13 LoanStats_2016Q4.csv
-rw-rw-r-- 1 juanan juanan  76M may  2 13:29 LoanStats_2017Q1.csv
-rw-rw-r-- 1 juanan juanan  83M may  2 13:47 LoanStats_2017Q2.csv
-rw-rw-r-- 1 juanan juanan  96M may  2 14:07 LoanStats_2017Q3.csv
-rw-rw-r-- 1 juanan juanan  93M may  2 14:27 LoanStats_2017Q4.csv
-rw-rw-r-- 1 juanan juanan  41M may  2 14:34 LoanStats3a.csv
-rw-rw-r-- 1 juanan juanan 155M may  2 15:04 LoanStats3b.csv
-rw-rw-r-- 1 juanan juanan 180M may  2 15:38 LoanStats3c.csv
-rw-rw-r-- 1 juanan juanan 319M may  2 16:37 LoanStats3d.csv


There are 12 files with a total disk size of 1.4 GB. All of them have been downloaded from [lending-club](https://www.lendingclub.com/info/download-data.action) web site. `LoanStats3a` file corresponds to the loans issued from 2007 to 2011, `LoanStats3b` from 2012 to 2013 and `LoanStats3c` and `LoanStats3d` to the years 2014 and 2015 respectively. From 2016, the loans have been registered by quarter (files named `LoanStats_YEARQUARTER.csv`)

Let us take a look at the first five observations to one of the csv files, for example, the loans issued in the last quarter of the last year:

In [9]:
!head -n 5 {raw_data_path}/LoanStats_2017Q4.csv

"id","member_id","loan_amnt","funded_amnt","funded_amnt_inv","term","int_rate","installment","grade","sub_grade","emp_title","emp_length","home_ownership","annual_inc","verification_status","issue_d","loan_status","pymnt_plan","url","desc","purpose","title","zip_code","addr_state","dti","delinq_2yrs","earliest_cr_line","inq_last_6mths","mths_since_last_delinq","mths_since_last_record","open_acc","pub_rec","revol_bal","revol_util","total_acc","initial_list_status","out_prncp","out_prncp_inv","total_pymnt","total_pymnt_inv","total_rec_prncp","total_rec_int","total_rec_late_fee","recoveries","collection_recovery_fee","last_pymnt_d","last_pymnt_amnt","next_pymnt_d","last_credit_pull_d","collections_12_mths_ex_med","mths_since_last_major_derog","policy_code","application_type","annual_inc_joint","dti_joint","verification_status_joint","acc_now_delinq","tot_coll_amt","tot_cur_bal","open_acc_6m","open_act_il","open_il_12m","open_il_24m","mths_since_rcnt_il","total_bal_il","il_util","open_rv_

Let us count the number of lines for each file using the shell:

In [10]:
list_dir = !ls {raw_data_path}

for file in list_dir:
    observations = !cat {raw_data_path}/{file} | wc -l
    print("{} has {} lines".format(file, observations[0]))

LoanStats_2016Q1.csv has 133892 lines
LoanStats_2016Q2.csv has 97859 lines
LoanStats_2016Q3.csv has 99125 lines
LoanStats_2016Q4.csv has 103551 lines
LoanStats_2017Q1.csv has 96784 lines
LoanStats_2017Q2.csv has 105456 lines
LoanStats_2017Q3.csv has 122706 lines
LoanStats_2017Q4.csv has 118653 lines
LoanStats3a.csv has 42543 lines
LoanStats3b.csv has 188186 lines
LoanStats3c.csv has 235634 lines
LoanStats3d.csv has 421100 lines


### 03 - Concatenate all files

We are going to concatenate all the previous files in just one dataframe:

In [11]:
list_ = []
for file in list_dir:
    full_path = os.path.join(raw_data_path, file)
    df = pd.read_csv(full_path, sep = ",")
    list_.append(df)
    
loans = pd.concat(list_)

Let us take a look at the first 10 rows and check up the shape of the dataframe:

In [14]:
loans.head(10)

"id"  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0   NaN        NaN      10000        10000            10000   60 months   
1   NaN        NaN      35000        35000            35000   60 months   
2   NaN        NaN      20000        20000            20000   60 months   
3   NaN        NaN      17475        17475            17475   60 months   
4   NaN        NaN       8000         8000             8000   36 months   
5   NaN        NaN      14400        14400            14400   36 months   
6   NaN        NaN      18000        18000            18000   60 months   
7   NaN        NaN       5800         5800             5800   36 months   
8   NaN        NaN      12500        12500            12500   60 months   
9   NaN        NaN       3000         3000             3000   36 months   

  int_rate  installment grade sub_grade       ...        \
0   19.53%       262.34     D        D5       ...         
1   20.75%       941.96     E        E2       ...         
2    9.16%       416.73     B        B2       ...         
3   11.47%       384.06     B        B5       ...         
4    9.16%       255.00     B        B2       ...         
5   10.75%       469.74     B        B4       ...         
6   11.99%       400.31     C        C1       ...         
7   11.47%       191.18     B        B5       ...         
8   14.46%       293.85     C        C4       ...         
9    7.39%        93.17     A        A4       ...         

  hardship_payoff_balance_amount hardship_last_payment_amount  \
0                            NaN                          NaN   
1                            NaN                          NaN   
2                            NaN                          NaN   
3                            NaN                          NaN   
4                            NaN                          NaN   
5                            NaN                          NaN   
6                            NaN                          NaN   
7                            NaN                          NaN   
8                            NaN                          NaN   
9                            NaN                          NaN   

  disbursement_method  debt_settlement_flag debt_settlement_flag_date  \
0                Cash                     N                       NaN   
1                Cash                     N                       NaN   
2                Cash                     N                       NaN   
3                Cash                     N                       NaN   
4                Cash                     N                       NaN   
5                Cash                     N                       NaN   
6                Cash                     N                       NaN   
7                Cash                     N                       NaN   
8                Cash                     N                       NaN   
9                Cash                     N                       NaN   

  settlement_status settlement_date settlement_amount  settlement_percentage  \
0               NaN             NaN               NaN                    NaN   
1               NaN             NaN               NaN                    NaN   
2               NaN             NaN               NaN                    NaN   
3               NaN             NaN               NaN                    NaN   
4               NaN             NaN               NaN                    NaN   
5               NaN             NaN               NaN                    NaN   
6               NaN             NaN               NaN                    NaN   
7               NaN             NaN               NaN                    NaN   
8               NaN             NaN               NaN                    NaN   
9               NaN             NaN               NaN                    NaN   

  settlement_term  
0             NaN  
1             NaN  
2             NaN  
3             NaN  
4             NaN  
5    

In [16]:
loans.shape

(1765451, 145)

We have 145 variables and 1765451 observations in the dataframe. But, as we can see looking at the head of the data, it seems that there are a lot of NaN columns we have to deal with.

### 04 - Select columns of interest

For selecting the columns of interest we are going to use the __data dictionary__ in excel format that lending-club provides in the web site.

In [19]:
columns_of_interest = ["funded_amnt_inv", "term", "issue_d","loan_status", "last_pymnt_d", 
                     "last_pymnt_amnt", "next_pymnt_d", "installment", "total_pymnt_inv",
                     "total_rec_late_fee", "int_rate", "sub_grade", "emp_title", "emp_length",
                     "annual_inc", "home_ownership", "title", "zip_code", "addr_state", "dti"]

In [22]:
loans = loans[columns_of_interest]

In [23]:
loans.head()

funded_amnt_inv        term   issue_d loan_status last_pymnt_d  \
0            10000   60 months  Mar-2016  Fully Paid     Nov-2016   
1            35000   60 months  Mar-2016  Fully Paid     Aug-2016   
2            20000   60 months  Mar-2016     Current     Feb-2018   
3            17475   60 months  Mar-2016     Current     Feb-2018   
4             8000   36 months  Mar-2016     Current     Feb-2018   

   last_pymnt_amnt next_pymnt_d  installment  total_pymnt_inv  \
0          9575.49          NaN       262.34         11127.83   
1           509.10          NaN       941.96         37226.47   
2           416.73     Mar-2018       416.73          9147.70   
3           384.06     Mar-2018       384.06          8432.61   
4           255.00     Mar-2018       255.00          5601.86   

   total_rec_late_fee int_rate sub_grade             emp_title emp_length  \
0                   0   19.53%        D5      lpn/charge nurse    4 years   
1                   0   20.75%        E2                Coiler    3 years   
2                   0    9.16%        B2  Reliability Engineer     1 year   
3                   0   11.47%        B5                   NaN        n/a   
4                   0    9.16%        B2            Technician  10+ years   

   annual_inc home_ownership               title zip_code addr_state    dti  
0       52000            OWN               Other    317xx         GA  15.00  
1       85000       MORTGAGE  Debt consolidation    144xx         NY  24.98  
2       77000       MORTGAGE    Home improvement    606xx         IL  13.75  
3       41682       MORTGAGE  Debt consolidation    796xx         TX  30.06  
4       72000           RENT  Debt consolidation    217xx         MD  22.63